eyelid-match-ops/src/primitives/yashe.rs

```diff
@@ -278,4 +346,116 @@ where
         res.truncate_to_canonical_form();
         res
     }
+
+    pub fn plaintext_add(&self, m1: Message<C>, m2: Message<C>) -> Message<C> {
+        let mut res = m1.m + m2.m;
+        for coeff in res.coeffs_mut() {
+            let mut coeff_res = C::coeff_as_u128(*coeff);
+            coeff_res %= C::t_as_u128();
+            *coeff = coeff_res.into();
+        }
+        res.truncate_to_canonical_form();
+        Message { m: res }
+    }
+
+    pub fn plaintext_mul(&self, m1: Message<C>, m2: Message<C>) -> Message<C> {
+        let mut res = m1.m * m2.m;
+        for coeff in res.coeffs_mut() {
+            let mut coeff_res: i128 = C::coeff_as_u128(*coeff) as i128;
+            // center lift mod q
+            if coeff_res > (C::modulus_minus_one_div_two_as_u128() as i128) {
+                coeff_res -= C::modulus_as_u128() as i128;
+            }
+            coeff_res %= C::t_as_u128() as i128;
+            if coeff_res < 0 {
+                coeff_res += C::t_as_u128() as i128;
+            }
+            *coeff = (coeff_res as u128).into();
+        }
+        res.truncate_to_canonical_form();
+        Message { m: res }
+    }
+
+    pub fn ciphertext_add(&self, c1: Ciphertext<C>, c2: Ciphertext<C>)-> Ciphertext<C> {
+        let c = c1.c + c2.c;
+        Ciphertext { c }
+    }
+
+    pub fn ciphertext_mul(&self, c1: Ciphertext<C>, c2: Ciphertext<C>)-> Ciphertext<C> {
+        let mut res = Poly::<C>::zero();
+        // lift to allow bignum coefficients (n * q * q would be enough, as in the C++ implementation)
+        let coeffs1 = c1.c.to_field_elements().unwrap();
+        let coeffs2 = c2.c.to_field_elements().unwrap();
+
+        let mut lifted_pol1 = DensePolynomial::<Fq79bn>::zero();
+        for i in 0..coeffs1.len() {
+            if i >= lifted_pol1.coeffs.len() {
+                lifted_pol1.coeffs.resize(i+1, C::coeff_as_u128(coeffs1[i]).into());
+            } else {
+                lifted_pol1[i] = C::coeff_as_u128(coeffs1[i]).into();
+            }
+        }
+        let mut lifted_pol2 = DensePolynomial::<Fq79bn>::zero();
+        for i in 0..coeffs2.len() {
+            if i >= lifted_pol2.coeffs.len() {
+                lifted_pol2.coeffs.resize(i+1, C::coeff_as_u128(coeffs2[i]).into());
+            } else {
+                lifted_pol2[i] = C::coeff_as_u128(coeffs2[i]).into();
+            }
+        }
+
+        // TODO: use more efficient version
+        let mut c = lifted_pol1.naive_mul(&lifted_pol2);
+        
+        // reduce by the cyclotomic polynomial
+        let mut i = C::MAX_POLY_DEGREE;
+        while i < c.coeffs.len() {
+            let q = i / C::MAX_POLY_DEGREE;
+            let r = i % C::MAX_POLY_DEGREE;
+
+            if q % 2 == 1 {
+                c.coeffs[r] = c.coeffs[r] - c.coeffs[i];
+            } else {
+                c.coeffs[r] = c.coeffs[r] + c.coeffs[i];
+            }
+            c.coeffs[i] = Fq79bn::zero();
+            i += 1;
+        }
```

Comment:
`Poly<OtherCoeff>::mul()`

Reply:
Yes, exactly.

Reply:
Turns out this had to be `mul` because it could change the size of the polynomial.